### PR TITLE
add a tutorial to 'My First Apps' for implementing SparkPI with Scalatra

### PR DESCRIPTION
**Pull request by reynoldsm88**
Docs are based on my project: https://github.com/reynoldsm88/tutorial-sparkpi-scala-scalatra.

Everything tested and working with Minishift:
minishift v1.20.0+53c500a
openshift v3.9.0+71543b2-33
kubernetes v1.9.1+a0ce1bc657

### CODE REVIEW

**Comment by elmiko** [comment]
hey @reynoldsm88 , thanks for the contribution =)

it might take a few days to get this reviewed, but i just wanted to let you know that we are taking a look and testing it out.

**Comment by reynoldsm88** [comment]
Thanks @elmiko. If there is anything you need me to do let me know and I'll be able to turn around the changes pretty quickly.

**Comment by reynoldsm88** [comment]
@elmiko BTW, I demoed this for my team today and I found an issue regarding using the scale parameter.

I'm going to track it as an issue here: https://github.com/reynoldsm88/tutorial-sparkpi-scala-scalatra/issues/1

When you are doing your testing for right now, can you just make sure to apply the ?scale=x. I'll provide the fix for using a defualt scale amount soon.

**Comment by reynoldsm88** [comment]
Issue with passing the "scale" param should be fixed in the demo repo with: https://github.com/reynoldsm88/tutorial-sparkpi-scala-scalatra/commit/ed67046c2e52b52f1582bb9d5c584abf08303aec

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    -p APP_MAIN_CLASS=io.radanalytics.examples.scalatra.sparkpi.Main \
+    -p APP_FILE=tutorial-sparkpi-scala-scalatra-assembly-0.0.1-SNAPSHOT.jar \
+    -p SBT_ARGS="clean assembly" \
+    -p APP_ARGS="-Xms1024M, -Xmx2048M, -XX:MaxMetaspace=1024M" \
+    -p OSHINKO_CLUSTER_NAME="spark"
+```
+
+Use `oc logs -f bc/tutorial-sparkpi-scala-scalatra` to tail the logs. The assembly task from SBT will be run, so it could take awhile. Once the application is ready you can expose it with `oc expose svc/tutorial-sparkpi-scala-scalatra`.
+
+Finally, you can get the IP address of your service endpoint. You can use this link in a web browser/REST client/curl command to submit a request and execute the Spark job.
```
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: I'll take care of that right now.

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    \_ target
+        \_scala-2.11
+            \_ tutorial-sparkpi-scala-scalatra-assembly-0.0.1-SNAPSHOT.jar
+
+....
+
+You can test it out by running `java -jar target/scala-2.11/tutorial-sparkpi-scala-scalatra-assembly-0.0.1-SNAPSHOT.jar`. However, it will fail if you do not have a Spark cluster running locally.
+
+== Deploy the Application to Openshift
+Assuming you have created a project, imported the https://radanalytics.io/resources.yaml[Radanalytics resources], and created a cluster called **`spark`**, all you should need to do to deploy the application is execute the following command:
```
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: Got it, no problem.

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 3]
```diff
@@ -0,0 +1,329 @@
+= SparkPi using Scala and Scalatra
+:page-layout: markdown
+:page-menu_template: menu_tutorial_application.html
+:page-menu_backurl: /my-first-radanalytics-app.html
+:page-menu_backtext: Back to My First RADanalytics Application
+
+== Building a Spark PI Microservice with Scalatra
+
+These instructions will help you to create a SparkPi microservice using https://www.scala-lang.org[Scala] and the http://scalatra.org[Scalatra] framework. In addition, this project will also demonstrate how to unit test your Spark jobs using the https://github.com/holdenk/spark-testing-base/wiki[Spark Testing Base] framework.
```
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: I noticed that last week. I think the issue has been resolved on their side and that it is reachable again.

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+// 5. Deployment of this artifact should be part of a CI/CD pipeline. Running the unit tests while building the "fat jar" is very expensive,
+//    therefore, don't do it during the "assembly" phase (which will be run on Openshift).
+test in assembly := {}
+
+// 6. Resolve any conflicts when merging into a "fat jar"
+assemblyMergeStrategy in assembly := {
+    case PathList( "META-INF", "MANIFEST.MF" ) => MergeStrategy.discard
+    case PathList( "reference.conf" ) => MergeStrategy.concat
+    //    case PathList( "META-INF", xs@_* ) => MergeStrategy.first
```
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: This code has been passed around between several projects before being distilled down into this. It's probably an artifact from another project. I'll remove it both in the docs as well as in the repo itself.

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+    \_ src
+        \_ main
+            \_ scala
+                \_ io
+                    \_ radanalytics
+                        \_
+                            \_ examples
+                                \_ scalatra
+                                    \_ sparkpi
+                                        \_ SparkPiTest.scala
```
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: There are a couple of things going on here. SBT derives a lot of it's functionality by following Maven conventions. Maven has more or less codified what a JVM derived language project should look like in the https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html.

The implication here has to do with the build artifacts. Anything placed underneath the `src/main` folder will wind up in the application that is produced at the end of the build. Anything in `src/test` will be compiled and added to the classpath for the testing phase only.

The end result is that the build artifacts do not include any test classes or test resources when they are deployed. This is a standard practice in the JVM community and should be familiar to any Scala/Java/Groovy developer.

[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: @elmiko Long story short re: ^^ my previous comment ^^ - this is the most realistic project structure one would have when developing a Scala application. This would also be true for Java applications as well as `src/main` and `src/test` are well codified conventions in organizing project source code.

**Comment by reynoldsm88** [comment]
@elmiko i'm going through the changes now. i should have them ready soon. having a problem with hotel wifi at the moment.

**Comment by elmiko** [comment]
no worries @reynoldsm88 =)

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: I see what I did there. I'll fix it to make it more clear. Also, good catch on the fact that I just copy and pasted main and didn't change it to test.

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+:page-layout: markdown
+:page-menu_template: menu_tutorial_application.html
+:page-menu_backurl: /my-first-radanalytics-app.html
+:page-menu_backtext: Back to My First RADanalytics Application
+
+== Building a Spark PI Microservice with Scalatra
+
+These instructions will help you to create a SparkPi microservice using https://www.scala-lang.org[Scala] and the http://scalatra.org[Scalatra] framework. In addition, this project will also demonstrate how to unit test your Spark jobs using the https://github.com/holdenk/spark-testing-base/wiki[Spark Testing Base] framework.
+
+You should already have the necessary prerequisites installed and configured, but if not please review the link:/my-first-radanalytics-app.html[instructions]. In addition, this tutorial will assume you have used the Oshinko Web UI to create a Spark cluster named **spark**. This will be used as the **OSHINKO_CLUSTER_NAME** parameter.
```
[the remark this answers is not in the document]
Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: I am changing the tutorial to conform to the other ones, so we'll use the ephemeral cluster. This will just be removed.

**Comment by elmiko** [comment]
hey @reynoldsm88 , i've been working through this tutorial again and i have a few more nits

1. when i run `sbt clean test assembly` locally i get a bunch of dependency errors, i'm guessing this a config issue on my side but i'd like to make sure the instructions are clear. not sure how we solve this but i am attaching the output of my build, maybe we should specify the sbt versions or something? 
[sbt.out.txt](https://github.com/radanalyticsio/radanalyticsio.github.io/files/2255218/sbt.out.txt)

1. the code blocks should use the highlighter that we have recently installed, you can see an example of how it's used [in this file](https://raw.githubusercontent.com/radanalyticsio/radanalyticsio.github.io/master/_howdoi/use-python-packages.adoc). basically just need to add a few headers at the top of the file and then change the markup around the code blocks.

1. in the deploy to openshift section, we should mention checking in the files (since this is supposed to be a follow-along type example) and then changing the `GIT_URI` parameter to their repo. the instructions themselves are fine, i just want to be sure that we getting users to run their code.

1. the link back to the tutorial page should be `link:/my-first-radanalytics-app.html#user[My First RADanalytics Application page`, this allows the site to be more portable without the hardcoded domain name.

1. i am able to run your upstream code, but when i follow the instructions in the tutorial and copy all the code into the files listed i am running into an error as the application launches. the build is fine, but i see this in the run logs:

````
spark-submit --class io.radanalytics.examples.scalatra.sparkpi.Main --master spark://cluster-0a4911:7077 --conf spark.driver.host=sparkpi-headless /opt/app-root/src/tutorial-sparkpi-scala-scalatra-assembly-0.0.1-SNAPSHOT.jar -Xms1024M, -Xmx2048M, -XX:MaxMetaspace=1024M
2018-08-02 21:38:16 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
java.lang.ClassNotFoundException: io.radanalytics.examples.scalatra.sparkpi.Main
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at org.apache.spark.util.Utils$.classForName(Utils.scala:235)
        at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:836)
        at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:197)
        at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:227)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:136)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
2018-08-02 21:38:17 INFO  ShutdownHookManager:54 - Shutdown hook called
````
beyond those changes, i think we just need to prepare for moving the code repository into the radanalyticsio organization. we should be very close to merging at that point.

thanks again for all the hard work =)

**Comment by elmiko** [comment]
for reference, here is my version of the app https://github.com/elmiko/sparkpi-scalatra

**Comment by reynoldsm88** [comment]
@elmiko - responding by number :)

1. I'll have a look at your build SBT file. Just out of curiousity, did you install SBT via a package manager like yum/dnf/homebrew or manually? I'm going to wipe out my Ivy cache and SBT settings and start from scratch. I'll get back to you.
2. That's an easy fix. Will do.
3. I can retool that section to fit that in.
4. Easy fix.
5. I'll check your repo out and try running it. That also looks like it could also be an SBT issue.

tl;dr - There are a few easy fixes to the docs. I'll check out your code and run it and see what's going on. Give me a day or two and I'll report back.

**Comment by elmiko** [comment]
sounds great, thanks for the quick response!

**Comment by reynoldsm88** [comment]
@elmiko Was there any other things I needed to do to fix up this repository? The only issue I had was that on my machine when I built the docs the syntax highlighting didn't render properly. I was using the Docker build method.

File: assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+tutorial-sparkpi-scala-scalatra (root)
+    \_ build.sbt
+    ...
+    \_ target
+        \_scala-2.11
+            \_ tutorial-sparkpi-scala-scalatra-assembly-0.0.1-SNAPSHOT.jar
+
+----
+
+This deployment can be tested loby running `java -jar target/scala-2.11/tutorial-sparkpi-scala-scalatra-assembly-0.0.1-SNAPSHOT.jar. The service should start, but will fail when servicing requests, unless you have a Spark cluster running locally.
```
Comment (elmiko) [inline assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: missing a closing quote on the "java -jar..." part

Comment (reynoldsm88) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: @elmiko sorry got busy with something else. I'll take care of this tonight.

Comment (elmiko) [reply assets/my-first-radanalytics-app/sparkpi-scala-scalatra.adoc]: no apologies necessary. i know the feeling all too well ;)

**Comment by reynoldsm88** [comment]
@elmiko I might have some SBT magic configured on my machine that help it work, I'll do some research into that.

**Comment by elmiko** [comment]
@reynoldsm88 i'm wondering if we shouldn't merge as it is, and then maybe update the language around building locally. mainly my concern is about adding some warnings for people who are running locally. what do you think?

**Comment by reynoldsm88** [comment]
@elmiko i did some investigation into your repo (https://github.com/elmiko/sparkpi-scalatra). There were some merge conflicts in there, but I pulled all my changes down and merged upstream master with no issue. I tested everything again today and it works.

Maybe check my comment on your repo https://github.com/elmiko/sparkpi-scalatra/pull/1#issuecomment-423332674. I ran locally and did see an innocuous Hadoop exception thrown, but everything builds locally for me and the tests pass. Seems the Hadoop error is immaterial.

Tested everything again on latest minishift and having no issues:
Mac OSX 10.13.6
minishift v1.24.0+8a904d0
openshift v3.10.0+e3465d0-44
kubernetes v1.10.0+b81c8f8

**Comment by reynoldsm88** [comment]
@elmiko At the same time, the local build was just to show thing from a developer's typical work flow. Build your app, run your tests, and then commit the code. Then the deployment to Openshift happens. If that part is not relevant we can just remove the section about the local build altogether.

If that's the case then, I need to find a way to make sure the tests run during the Openshift build. They don't now because they were consuming so much memory that they blew the pod up. I'm limited to my laptop Openshift environment so there is only so far I can beaf them up.

**Comment by elmiko** [comment]
ok, thanks for the explanation.

i don't think we should have the tests in the openshift build, but i guess my big concern is support for people who might to try run locally and have issues. i think the instructions and discussion are well placed, i'm just not sure about the best way to proceed since i ran into issues trying to build locally and i can imagine others running into that as well.

maybe just best to merge as it is and see if any issues arise.

**Comment by elmiko** [comment]
thanks for the updates, and all the hard work. i think this is ready to publish =)
